nexus/db-queries/src/db/datastore/inventory.rs

```diff
@@ -0,0 +1,914 @@
+        //   one.
+        //
+        // - There could be fewer collections in the database, early in the
+        //   system's lifetime (before we've accumulated `nkeep` of them).
+        //
+        // - There could be many more collections in the database, if something
+        //   has gone wrong and we've fallen behind in our cleanup.
+        //
+        // - Due to transient errors during the collection process, it's
+        //   possible that a collection is known to be potentially incomplete.
+        //   We can tell this because it has rows in `inv_collection_errors`.
+        //   (It's possible that a collection can be incomplete with zero
+        //   errors, but we can't know that here and so we can't do anything
+        //   about it.)
+        //
+        // Goals:
+        //
+        // - When this function returns without error, there were at most
+        //   `nkeep` collections in the database.
+        //
+        // - If we have to remove any collections, we want to start from the
+        //   oldest ones.  (We want to maintain a window of the last `nkeep`,
+        //   not the first `nkeep - 1` from the beginning of time plus the most
+        //   recent one.)
+        //
+        // - We want to avoid removing the last collection that had zero errors.
+        //   (If we weren't careful, we might do this if there were `nkeep`
+        //   collections with errors that were newer than the last complete
+        //   collection.)
+        //
+        // Here's the plan:
+        //
+        // - Select from the database the `nkeep + 1` oldest collections and the
+        //   number of errors associated with each one.
+        //
+        // - If we got fewer than `nkeep + 1` back, we're done.  We shouldn't
+        //   prune anything.
+        //
+        // - Otherwise, if the oldest collection is the only complete one,
+        //   remove the next-oldest collection and go back to the top (repeat).
+        //
+        // - Otherwise, remove the oldest collection and go back to the top
+        //   (repeat).
+        //
+        // This seems surprisingly complicated.  It's designed to meet the above
+        // goals.
+        //
+        // Is this going to work if multiple Nexuses are doing it concurrently?
+        // This cannot remove the last complete collection because a given Nexus
+        // will only remove a complete collection if it has seen a newer
+        // complete one.  This cannot result in keeping fewer than "nkeep"
+        // collections because any Nexus will only remove a collection if there
+        // are "nkeep" newer ones.  In both of these cases, another Nexus might
+        // remove one of the ones that the first Nexus was counting on keeping,
+        // but only if there was a newer one to replace it.
+
+        opctx.authorize(authz::Action::Modify, &authz::INVENTORY).await?;
+
+        loop {
```

Comment:
I think this is a promising approach.  But I'm a little worried it'll take a while to make this real (first real SQL, then real Diesel code), and also that when we do, we may find the database winds up doing a table scan (or rejecting the SQL because we've configured it to disallow that).  As an example: one of the stated assumptions is that the number of collections here _could_ be huge.  In that case, the highest-level subquery here will produce only 3 rows, but the query itself will be trying to select any collections _not_ in that set, which will in turn return a very large result set.  So we'll a `LIMIT` there.  We also want to start with the oldest ones, so we'll want an `ORDER BY` timestamp.  At this point, there are enough variables here that I'm not sure what query plan Cockroach will use.  I think ideal would be to do the subquery, then scan the index (by timestamp) and just skip over any rows that are in the subquery and stop when we reach the limit.  I hope it will do that but it's hard to be sure until we do that work.

I think this is probably all solvable (or else we'll find out why it's not), but if the current code is at least correct and not pathological, I'd rather defer this than spend the time now to work all this out.